zrml/authorized/src/lib.rs

```diff
@@ -60,36 +63,63 @@ mod pallet {
         <<T as Config>::MarketCommons as MarketCommonsPalletApi>::MarketId;
     pub(crate) type MomentOf<T> = <<T as Config>::MarketCommons as MarketCommonsPalletApi>::Moment;
 
+    pub type CacheSize = ConstU32<64>;
+
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         /// Overwrites already provided outcomes for the same market and account.
         #[frame_support::transactional]
-        #[pallet::weight(T::WeightInfo::authorize_market_outcome())]
+        #[pallet::weight(T::WeightInfo::authorize_market_outcome(
+            CacheSize::get(),
+            CacheSize::get()
+        ))]
         pub fn authorize_market_outcome(
             origin: OriginFor<T>,
             market_id: MarketIdOf<T>,
             outcome: OutcomeReport,
-        ) -> DispatchResult {
+        ) -> DispatchResultWithPostInfo {
             T::AuthorizedDisputeResolutionOrigin::ensure_origin(origin)?;
             let market = T::MarketCommons::market(&market_id)?;
             ensure!(market.status == MarketStatus::Disputed, Error::<T>::MarketIsNotDisputed);
             ensure!(market.matches_outcome_report(&outcome), Error::<T>::OutcomeMismatch);
-            match market.dispute_mechanism {
-                MarketDisputeMechanism::Authorized => {
-                    AuthorizedOutcomeReports::<T>::insert(market_id, outcome);
-                    Ok(())
-                }
-                _ => Err(Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized.into()),
-            }
+            ensure!(
+                market.dispute_mechanism == MarketDisputeMechanism::Authorized,
+                Error::<T>::MarketDoesNotHaveDisputeMechanismAuthorized
+            );
+
```

Comment:
I renamed the function `is_fail` to `is_expired`. Your point stays the same, but it's maybe not a huge problem after that. Sure, we could reject authority reports after that, but I guess even after expiration, it should be still allowed to authorize. The authority was in this case lucky, that nobody cared to resolve this market, but everyone had the optionality to do so. I like to have more optionalities here rather than to add another constraint.

Reply:
I agree: It's not a huge problem. But I think we should have clear rules. Why would an MDM expire, and then suddently no longer be expired? I think the spec should have a note saying "If this returns `true` when called, it must always return `true` in the future."

There are also technical reasons. When the MDM expires, that passes control over the dispute back to `prediction-markest`, which can then initiate a global dispute. When a global dispute is initiated, the MDM doesn't know this, correct? There's no way for it know. And control cannot (and should not) be returned to the MDM.

But that's a problem: After the global dispute is initiated, the authority wakes up from its beauty sleep and submits its report, and `zrml_authorized` calls `add_auto_resolve`. So now we have a resolution queued, but there's also a global dispute going on. That should never happen.

Reply:
Ok, got you! 🥇 nice arguments

fb48a7ef794276819357a6f735b379c5e676e9fb